### ISSUE TITLE
Add late physics process

### ISSUE THREAD

**Issue by KeyboardDanni** [godotengine/godot-proposals#6795]
### Describe the project you are working on

2D sidescroller

### Describe the problem or limitation you are having in your project

A continuation of https://github.com/godotengine/godot/issues/19228 which oddly enough, has not seen a feature proposal here yet.

Right now Godot has several issues with things happening "one frame too late" because they aren't getting processed in the optimal order each frame. In particular, anything that should occur as a result of physics engine updates only gets a chance to run in the same frame if it's done in idle process, but doing this has the potential to make the playsim non-deterministic.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

It would be great if Godot had a "late physics process" - i.e. an update that occurs after `_physics_process` but before `_process`, and only on physics frames. This would allow things to be processed in the correct order:

- Physics process: character movement logic
- Physics engine update: simulate moving objects
- Late physics process: update camera, animations
- Idle process: render everything

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

Introduce a new function, `_late_physics_process`, to go alongside `_physics_process` and `_process`. Nodes would be able to implement this new function for anything that should happen after the physics engine step, but only on physics frames.

Additionally, built-in nodes like Camera and AnimationPlayer would have a new "Late Physics" process mode which would use `_late_physics_process`.

My only concern with this change is that it might cause extra object iteration overhead. So it might be a good idea to maintain a list of objects that opt into late physics process and only iterate through those.

### If this enhancement will not be used often, can it be worked around with a few lines of script?

Workarounds exist but they're kinda iffy, and since this issue affects built-in nodes too, you can't really script around the problem for those.

### Is there a reason why this should be core and not an add-on in the asset library?

Issue affects core nodes and the core engine loop, so can't really be fixed with an add-on.

**Comment by KoBeWi** [comment]
I think you can use `process_priority` for this. According to the documentation, it should also affect built-in nodes.

**Comment by KeyboardDanni** [comment]
My understanding is `process_priority` only affects the order of execution of `_physics_process` in relation to other nodes, but can't be used to make a node get processed after the physics engine update. For example, when using a Camera2D to track a CharacterBody2D, no matter what I set the `process_priority` to, the camera always lags behind the target object by a frame if the camera's process mode is set to "Physics". The issue doesn't occur if the process mode is "Idle". But I'd like to have everything use "Physics" if possible. Using "Idle" for Camera2D is probably innocuous, but having an AnimationPlayer process outside of physics could desync the playsim if any logic is dependent on the current animation frame.

**Comment by smix8** [comment]
That issue that you describe is the common misconception of using physics objects for something unrelated to physics e.g. visual updates or rendering.

If Godot had a default physics tickrate of e.g. 10 people would immediately spot how wrong using physics for a camera or a visual animation is but since the default is 60 it goes unnoticed for long until the (daily) complains popup. Even with physics interpolation this is never really solved, it is just the wrong usage.

That late update idea will not solve your issue because physics runs at 0 - 8 ticks each rendered frame. While users will always only see the new frame rendered there is no guaranteed that anything physics related has been updated properly or updated at all in that frame when it comes to visuals.

So TLDR don't use physics_process for visual updates.
That is the wrong use of physics_process when you want smooth and accurate camera and animations.

**Comment by KeyboardDanni** [comment]
> That issue that you describe is the common misconception of using physics objects for something unrelated to physics e.g. visual updates or rendering.

Camera updates are not "unrelated to physics", because if the camera position is used to handle activation/deactivation of objects, you may end up with objects spawning on different physics frames each time you play, which can (will) desync the playsim. Animation updates are not "unrelated to physics" because the same thing can happen. Attaching an object to a player's arm, for example, may cause said object to move to a different position on the wrong physics frame, and cause a playsim desync.

> If Godot had a default physics tickrate of e.g. 10, people would immediately spot how wrong using physics for a camera or a visual animation is but since the default physics tickrate is 60 it goes unnoticed for long until the complains popup. Even with physics interpolation this is never really solved, it is just the wrong usage.

SNES, Genesis games didn't use a tickrate of 10, they used a rate of 60. Nobody uses a physics rate of 10 unless they're either targeting very old PC hardware or they're making something esoteric on purpose. Even then, it's more typical to use a tickrate of 30 or even 20.

> That late update idea will not solve your issue because physics runs at 0 - 8 ticks each rendered frame. While users will always only see the new frame rendered there is no guaranteed that anything physics related has been updated properly or updated at all in that frame when it comes to visuals.

I'm not sure how this is an issue here. If physics runs twice a frame, we can just run `_physics_process` followed by `_late_physics_process`, then do the same thing again before the idle process.

> So TLDR don't use physics_process for visual updates. That is the wrong use of physics_process when you want smooth and accurate camera and animations, especially on high frame rate with a lower physics tickrate. Make your visual objects follow / react smoothly to your physics objects, not push the physics results to your visuals.

If anything, the camera jitter I've seen in Godot has largely been due to 1. improper combined use of physics and idle (updating positions on physics and camera on idle), and 2. bad framepacing in Godot (which should be fixed by using physics interpolation, waiting to do at least one physics update before rendering, and using the native platform swapchain, not by papering over it with idle process).

**Comment by RedMser** [comment]
They don't literally mean 10 fps physics update. Rather, having a lower physics rate than process rate. You would have the same issue on 60 fps physics and 144 Hz rendering rate, but most people have only 60 Hz monitors and thus never run into this.

**Comment by Calinou** [comment]
>  you may end up with objects spawning on different physics frames each time you play, which can (will) desync the playsim

Godot's physics engine is not deterministic, and neither is Bullet in 3.x.

**Comment by KeyboardDanni** [comment]
> They don't literally mean 10 fps physics update.

Oops, my bad. I do maintain that you'd have the same issue with 60 FPS at 144hz, as you mentioned.

> most people have only 60 Hz monitors and thus never run into this.

Not necessarily. A few things: first, not all 60hz configurations are equal. Back in the CRT days, different resolutions all had different ideas of what 60hz was exactly. For example, 800x600 was actually closer to 61hz, from what I recall. This is still somewhat true today, though it will vary based on your display and graphics hardware. You also have the NTSC standard which is "not quite" 60hz but rather 59.94hz. Here's the list of refresh rates I get on my machine:

![2023-05-03 18_57_23-Settings](https://user-images.githubusercontent.com/34800072/236068694-039fe74c-fcb1-45ad-86c8-10844528a126.png)

Second, and probably more important, is that framepacing issues can happen even at 60hz. If the engine decides not to process a physics frame, but then does a Vsync, you end up with zero physics frames and one idle frame. Because of Vsync, you end up waiting 16ms even if the next physics frame was only going to be in 2ms. Then physics has to catch up, so next frame you get two physics frames and one idle frame. This is the single biggest cause of framepacing issues I've seen in games (that isn't related to the compositor). Workarounds exist, such as framepacing extensions that let you query timings, plus triple buffering, though the latter may come at the cost of increased input lag. Not to mention if you OS doesn't have direct scanout, it'll probably go through the compositor, which means another potential source of framepacing woes.

> Godot's physics engine is not deterministic, and neither is Bullet in 3.x. While we intend to make GodotPhysics slightly more deterministic to allow for better client-side prediction, there are no plans to make it 100% deterministic in a way that would be reliable for storing input-based replays (it's recommended to store transforms instead). Therefore, I wouldn't rely on any form of "incidentical" determinism that may occur.

I believe I know what causes this (acceleration structure resulting in objects being ordered differently), which is unfortunate, but at least it's not related to timing. If there's a bit more randomness in a complex sim, so be it I guess. My big concern however is two users getting different results based solely on timing issues (different refresh rates, different compositor configuration, different OS, game hitched in a different spot for one user). I've seen how games can break when physics and idle get out of sync and it often isn't pretty.

My focus is less on the "old" input replay netcode from the Doom era. Which, you're right, it's not great. Not too long ago I tried to play a netgame on ZDoom with friends and it had Problems.

No, I'm personally more interested in stuff like precision platformers, speedrunning, that sort of thing. If a game like Celeste had issues with non-deterministic physics, I think people would pick up on that pretty quick.

> With physics interpolation in place, this would look just fine (and better than an interpolated 10 FPS camera following a 60 FPS object). This is especially the case when mouselook is involved with a third-person character. Games like Heretic 2 or Anachronox come to mind here, as they inherit from Quake 2's 10 Hz tickrate.

This would probably depend on when the camera is getting repositioned. In some cases it might help, but not completely eliminate the jitter issue. But as long as the target and camera are getting repositioned on the physics tic and are both being interpolated the same way, it should be fine in theory.

For reference, here is a watered down version of what I'm currently using for the camera:
```
func _physics_process(_delta):
    if is_instance_valid(target):
        position.x = floor(target.position.x);
        position.y = floor(target.position.y);
```

If "Process Callback" is set to "Idle", it works fine. If it's set to "Physics", it lags behind by a frame. But I feel like this setup only really works as well as it does because the camera's transform is changed on the physics tic, and the idle process happens to be after the physics engine update.

**Comment by wagnerfs** [comment]
Just dropping to add my 2 cents here!
I noticed something after someone ran my game on a slow machine, at like, 20 FPS, all my bones logics like jiggle physics and bone look at got all wonky, after doing some research I realized the issue is because `_physcis_process` where the code is supposed to run, happens BEFORE the animation updates the bones chain, so by the time it reaches rendering, there will be a gap between bone positioning and structure.

Godot has a signal called `frame_pre_draw` from RenderingServer, which's basically a post physics/animation and right before everything is rendered, by moving my code there all the bone positioning issues stopped, so in short, godot does have a late update method you can connect, but you'd still need to limit it to the same `physics_process` frame rate, and also get the delta yourself as well.

So yeah, I do agree we need a proper `_late_physics_process` with the delta included and preferebly right before the render code for comodity instead of writing everything I mentioned by ourselves in gdscript.

**Comment by KoBeWi** [comment]
If you do `call_deferred()` in `_physics_process()`, it will run after everything else.

**Comment by wagnerfs** [comment]
> If you do `call_deferred()` in `_physics_process()`, it will run after everything else.

Pretty sure that only queues the call for the next idle frame, which has nothing to do with execution order between process and animations.

**Comment by KoBeWi** [comment]
It's a common misconception about deferred calls. The docs were updated recently to clear that.

**Comment by Theome** [comment]
I have created a pull request that implements this.

> It's a common misconception about deferred calls. The docs were updated recently to clear that.

I think the docs could be improved. Right now the [docs state](https://docs.godotengine.org/en/stable/classes/class_callable.html):

>  Deferred calls are processed at idle time. Idle time happens mainly at the end of process and physics frames.

So one might assume that in the deferred call the physics engine step already happened, but that's not the case.

**Comment by AThousandShips** [comment]
It says:
> Deferred calls are processed at idle time. Idle time happens mainly at the end of process and physics frames. In it, deferred calls will be run until there are none left, which means you can defer calls from other deferred calls and they'll still be run in the current idle time cycle. This means you should not call a method deferred from itself (or from a method called by it), as this causes infinite recursion the same way as if you had called the method directly.

**Comment by Summersay415** [comment]
I think `physics_process_priority` already solved this?

**Comment by jitspoe** [comment]
Thanks for taking a shot at this.  I took a look at the diff (I assume it's this one: https://github.com/godotengine/godot/compare/master...Theome:godot:late-physics).  Is there any functional difference between using _late_physics_process(), and just:

```gdscript
# Camera script or whatever that needs to execute later

func _physics_process(deta):
    call_deferred("_late_physics_process", delta)

func _late_physics_process(delta):
    global_position = target_position # or whatever
```

?

**Comment by Theome** [comment]
>  Is there any functional difference between using _late_physics_process(), and just:
```
# Camera script or whatever that needs to execute later

func _physics_process(deta):
    call_deferred("_late_physics_process", delta)

func _late_physics_process(delta):
    global_position = target_position # or whatever
?
```

It's not the same. `call_deferred` is evaluated *before* the physics engine step.

**Comment by AThousandShips** [comment]
> It's not the same. call_deferred is evaluated before the physics engine step.

No it's evaluated at the end of `_process` and at the end of `_physics_process`, so if it is called during `_physics_process` it should be evaluated at the end of the physics step (haven't verified but the code does the correct thing in `SceneTree`)

**Comment by KoBeWi** [comment]
Relevant code:
https://github.com/godotengine/godot/blob/172fc62c803276f672967d296c1ca551884a068f/scene/main/scene_tree.cpp#L525-L543
`_physics_process()` happens at 541
Deferred calls happen at 534

Which is the physics step?

**Comment by Theome** [comment]
> Which is the physics step?

Here, outside of the call to `SceneTree`.

https://github.com/godotengine/godot/blob/172fc62c803276f672967d296c1ca551884a068f/main/main.cpp#L4472-L4479

**Comment by wagnerfs** [comment]
I feel like the issue currently is that the deferred calls still happen at the end of the normal physics process loop and what's being proposed is a function call AFTER all the internal physics updates have been processed but before the rendering is applied.

**Comment by Theome** [comment]
> > ... what's being proposed is a function call AFTER all the internal physics updates have been processed but before the rendering is applied.
> 
> That already exists, it is called the `process()` callback function. You want to do something just a moment before rendering commits the frame? You do it there.

That doesn't work if I need the call exactly once per physics frame.

**Comment by KeyboardDanni** [comment]
Not to mention, `process()` will result in non-deterministic behavior, as `physics_process()` may happen multiple times in a given iteration of the main engine loop, but `process()` is only called once. The whole point of `physics_process()` is keeping your code deterministic so you don't suddenly have different behavior if the frametime fluctuates (and it's pretty much *guaranteed* to fluctuate *somehow*).

**Comment by KeyboardDanni** [comment]
I will say, however, that the problem I described in this [comment](https://github.com/godotengine/godot-proposals/issues/6795#issuecomment-1533873964) isn't really an issue anymore, as I have since identified the cause of the one frame camera delay and [ported the fix from 3.x to 4.x](https://github.com/godotengine/godot/pull/84465). Now you can get a perfectly frame-synced camera just by changing the Camera2D's `process_physics_priority` to a positive value. The result looks [very nice](https://github.com/godotengine/godot-proposals/discussions/9256#discussioncomment-12663594).

(Aside: I still don't understand why the fix was made in 3.x first and left out of 4.x for over two whole years...)

**Comment by smix8** [comment]
The physics_process() markes both the begin of the new physics step as well als marks the end of the prev physics step. It is a stepped process after all, it thinks and acts in steps, not in frames. The physics is locked inbetween, especially while threaded. Trying to unlock it again, or trying to do a double sync like proposed here, is a sign for an XY problem where people try to push for core hacks because "it is easier" instead of fixing the actual issues where they are caused.

Those issues are usually related to some nodes not having the required api or callbacks exposed (node issue, not physics issue), or the user actually misunderstanding / misusing the system. E.g. moving frame related camera positions with physics steps and expecting it to run smooth. No wonder that backporting a transform fix "fixes" such issues because all it does is fixing the original user mistake of using the wrong process mode to update things, and that should have been the fix all along instead of hacking physics updates.

**Comment by KeyboardDanni** [comment]
> The physics_process() markes both the begin of the new physics step as well als marks the end of the prev physics step. It is a stepped process after all, it thinks and acts in steps, not in frames.

Yes. That's what I said already, and that's why `process()` is not a substitute for `physics_process()`.

> The physics is locked inbetween, especially while threaded. Trying to unlock it again, or trying to do a double sync like proposed here, is a sign for an XY problem where people try to push for core hacks because "it is easier" instead of fixing the actual issues where they are caused.

I never advocated for unlocking physics. Only for the option to be able to run scripts after the built-in physics step, but still have it happen alongside `physics_process()` so everything remains deterministic.

> Those issues are usually related to some nodes not having the required api or callbacks exposed (node issue, not physics issue), or the user actually misunderstanding / misusing the system. E.g. moving frame related camera positions with physics steps and expecting it to run smooth.

But if everything happens in the physics step it *will* be smooth. Moving some things in idle step and others in physics step is one of the biggest causes of visual jitter.

> No wonder that backporting a transform fix "fixes" such issues because all it does is fixing the original user mistake of using the wrong process mode to update things, and that should have been the fix all along instead of hacking physics updates.

The bug was that the camera's internal scrolling book-keeping wasn't getting updated when it was supposed to. The original logic in core was flawed. That PR wasn't a hack, that was just fixing the logic so it worked correctly.

> The solution should be to fix the actual issues and not hacking core.

The actual issue with the camera was in core. So I fixed it in core.

> So it would be good to have a list of very specific issues that users face with current nodes and classes that they think they can only fix by hacking into the core physics loop. That is always the wrong fix. We should allow users to solve the problems somehow else, either by improving the nodes (e.g. adding callbacks for specific node classes) or directing them to the correct use (e.g. explain why using physics process for certain things is bad idea).

I've been tinkering with gamedev since the Klik & Play days. I have always done everything in physics process and it's worked fine. The only problem with doing things in physics process is you can't take advantage of higher refresh rates, and physics interpolation addresses that. Any problems that might arise are largely due to updates happening out of sync because they're idle instead of physics (that includes the engine's current behavior of rendering a new frame even if there hasn't been a physics step update).

Please take the time to get acquainted with the issue at hand. I don't like getting passive-aggressively blamed due to "user error" which isn't actually user error, because I am *not* new to game development and I've done this all the time. This sort of attitude is only going to discourage people from contributing to the project.

**Comment by Theome** [comment]
I ran into this issue again in my project and found a workaround for me. I still believe that the proposed callback would be useful, so I have outlined my understanding of the relevant parts of Godot's main loop. I hope this can help clarify what properties the proposed callback would have. Please correct me if I have missed anything:

```
k = calculate_required_physics_steps() // (1)
for (i = 0; i < k; i++) {
  current_physics_step += 1; // (2)
  physics_process() // (3)
  flush_deferred_queue() // (4)
  physics_server_step() // (5)
  // late_physics_process() // (6)
}
process() // (7)
flush_deferred_queue() // (8)
render_frame() // (9)
```

With "physics step" I will refer to one iteration of the loop above, i.e. steps (2)-(6). 

Walking through it:

(1) Determine the number of physics steps that need to be processed. This can be any number between 0 and 8.

(2) Increase the physics step counter. This is the start of a new physics step.

(3) Call all `physics_process` methods implemented by the user. The order is determined by node tree order and physics priority. This is the place where all calls to `move_and_collide()` etc. should happen.

(4) Flush the queue with all deferred calls that might have been enqueued in the `physics_process` methods.

(5) The physics server does a step

(6) Here's the proposed new callback

(7) All physics steps have been fully processed. Here, all `process` methods implemented by the user are called.

(8) Flush the queue with all deferred calls that might have been enqueued in the `process` methods.

(9) The frame is rendered.


The suggestion is to have a callback at position (6). At this point, the physics server has just finished a step, and if it's the last iteration, the physics state is what we're actually going to see on screen in the following render call (9). In a sense, this is the "definite state" for the current physics step. 

Now let's consider I want to do some specific thing "X" at point (6), and look at some of the suggestions that came up in this thread:

"Do X in a deferred call"

This is too early: The deferred calls happen in (4), but the physics server step hasn't happened yet.

"Do X in `process`"

This doesn't work if X must happen exactly once per physics step. For each call to `process`, zero, one, or several physics steps can happen.

"Adjust physics priorities and do X as late as possible in `physics_process`"

This is too early: We only reorder the calls within (3), but the physics server step hasn't happened yet.

"Adjust physics priorities and do X as early as possible in `physics_process`"

While we do have access to the "definite state" of a physics step at this point, it's the state of the **previous** step. That step has already been rendered. Also, X now happens one physics step too late, and that additional latency might be too much.

To summarize the properties of the suggested new callback:

- It happens *exactly once* per physics step
- It happens *after* the physics server step, i.e. allows access to the "definite state"
- It happens *immediately* after the physics server step, without a significant delay. 
- It allows access to game objects after the physics server step, and *before they are being rendered*.

My personal use case is the following: My game is a fast paced multiplayer action game, and I use custom networking. My networking/synchronization model is based on the "definite state" of each physics step, and I need to access it with as little delay as possible. The suggested callback would be the perfect place for that. My current workaround is to use `CharacterBody3D` objects instead of `RigidBody` objects, so that my game logic doesn't depend on the physics server step at all.

If I read the feedback on my PR correctly, my suggested implementation negatively impacts the performance. I wasn't aware of this, as I don't understand the physics server design well enough. My main goal with the PR was to provide a concrete proposal to anchor the discussion, not to push for this specific implementation. If it's fundamentally flawed, I'm happy to close the PR. I'm still wondering, is there a viable way to get a callback at this specific point in the loop?

**Comment by Kaleb-Reid** [comment]
This is also something that I need for my project where I don't think there is any workaround for.

I essentially need to emulate a child node following a parent node when the parent's transform changes, but the nodes are not allowed to be directly related (parent cannot be an ancestor of child because reparenting is not an option since the child is controlled by a multiplayer peer and the child can start following a different parent at any time.

- Doing it in ```_physics_process``` causes physics glitches if the bodies are moving fast enough (not very fast) due to being behind 1+ physics frames.
- Doing it in ```_process``` would work if I wasn't using physics interpolation but I am and I think the transform update needs to happen before the interpolation state is updated.
- Doing it in a deferred call or changing the physics priority doesn't really help me when I'm relying on rigid body states and tweens using physics processing which still update / happen after.

**Comment by v1993** [comment]
@Kaleb-Reid You may want to either directly use or take inspiration from RemoteTransform2D/RemoteTransform3D nodes.

**Comment by Kaleb-Reid** [comment]
@v1993 I did consider using them but the problem is that using a RemoteTransform3D or applying transforms during ```NOTIFICATION_TRANSFORM_CHANGED``` (same as RemoteTransform3D and normal scene tree transform propagation) causes rigid bodies to stop responding to physics for some reason and I'm not really sure of the reason why.

Now that I'm thinking of it again I guess there is a possibility that it would also do that in the proposed ```_late_physics_process``` but I havent tried it yet.

EDIT: I realized that I needed to call ```force_update_transform()``` in order to make rigid bodies work while transforming during ```NOTIFICATION_TRANSFORM_CHANGED```. ```_late_physics_process``` would still be useful for doing physics queries more precisely but does remove the point of doing multi-threaded physics.

**Comment by Kaleb-Reid** [comment]
Although I don't really need this anymore I did some more thinking about it. After looking through the linked pr, I realized that it's a bad idea to sync the physics thread with the main thread again after calling ```step()``` because that defeats the purpose of running the physics server on another thread. Instead of the main loop continuing while the physics server steps, the main loop waits for the physics server to step while doing nothing itself.

I guess that the original idea would make sense if the physics server is running on the same thread and ```_late_physics_process()``` could be disabled otherwise. But another solution which could work in both cases without getting really hacky could be to instead run callbacks from the physics thread at the end of the step.

I would imagine the API would include something like ```PhysicsServer*D.add_post_step_callback(callable: Callable)``` and ```PhysicsServer*D.remove_post_step_callback(callable: Callable)```. If using multi-threaded physics, anything with the physics state that needs to be done such as queries can happen in the callback and anything else can be called deferred on the main thread. If not using multi-threaded physics then you don't need to worry about which thread you are calling from.
